lib/driver-cli.js

```diff
@@ -0,0 +1,249 @@
+import _ from 'lodash';
+import path from 'path';
+import { KNOWN_DRIVERS } from './drivers';
+import DriverConfig, { INSTALL_TYPE_NPM, INSTALL_TYPE_GIT, INSTALL_TYPE_GITHUB,
+                       INSTALL_TYPE_LOCAL } from './driver-config';
+import NPM from './npm';
+import { errAndQuit, log, spinWith, JSON_SPACES } from './cli-utils';
+import { util, fs } from 'appium-support';
+
+async function runDriverCommand (args) {
+  // TODO driver config file should be locked while any of these commands are
+  // running to prevent weird situations
+  let jsonResult = null;
+  const {json, appiumHome} = args;
+  const logFn = (msg) => log(json, msg);
+  const config = new DriverConfig(appiumHome, logFn);
+  const cmd = new DriverCommand({config, json});
+  try {
+    await config.read();
+    jsonResult = await cmd.execute(args);
+  } catch (err) {
+    errAndQuit(json, err);
+  }
+
+  if (json) {
+    console.log(JSON.stringify(jsonResult, null, JSON_SPACES)); // eslint-disable-line no-console
+  }
+}
+
+class DriverCommand {
+
+  constructor ({config, json}) {
+    this.config = config;
+    this.json = json;
+    this.npm = new NPM(this.config.appiumHome);
+  }
+
+  async execute (args) {
+    if (!this[args.driverCommand]) {
+      throw new Error(`Cannot handle driver command ${args.driverCommand}`);
+    }
+    return await this[args.driverCommand](args);
+  }
+
+  async list ({showInstalled, showUpdates}) {
+    const lsMsg = `Listing ${showInstalled ? 'installed' : 'available'} drivers`;
+    const drivers = {};
+
+    for (const name of Object.keys(this.config.installedDrivers)) {
+      drivers[name] = this.config.installedDrivers[name];
+      drivers[name].installed = true;
+    }
+
+    if (!showInstalled) {
+      for (const name of Object.keys(KNOWN_DRIVERS)) {
+        if (!drivers[name]) {
+          drivers[name] = {
+            pkgName: KNOWN_DRIVERS[name],
+            installed: false,
+          };
+        }
+      }
+    }
+
+    await spinWith(this.json, lsMsg, async () => {
+      if (showUpdates) {
```

Comment:
`if (!showUpdates) return`

it is easier to read the code if the nesting level is low